lib/core/src/Cardano/Wallet/Primitive/Types/Tx/Gen.hs

```diff
@@ -0,0 +1,109 @@
+{-# LANGUAGE DataKinds #-}
+
+module Cardano.Wallet.Primitive.Types.Tx.Gen
+    ( genTxHashSmallRange
+    , genTxIndexSmallRange
+    , genTxInSmallRange
+    , genTxOutSmallRange
+    , shrinkTxHashSmallRange
+    , shrinkTxIndexSmallRange
+    , shrinkTxInSmallRange
+    , shrinkTxOutSmallRange
+    )
+    where
+
+import Prelude
+
+import Cardano.Wallet.Primitive.Types.Address.Gen
+    ( genAddressSmallRange, shrinkAddressSmallRange )
+import Cardano.Wallet.Primitive.Types.Hash
+    ( Hash (..) )
+import Cardano.Wallet.Primitive.Types.TokenBundle.Gen
+    ( genTokenBundleSmallRange, shrinkTokenBundleSmallRange )
+import Cardano.Wallet.Primitive.Types.Tx
+    ( TxIn (..), TxOut (..) )
+import Data.Either
+    ( fromRight )
+import Data.Text.Class
+    ( FromText (..) )
+import Data.Word
+    ( Word32 )
+import Test.QuickCheck
+    ( Gen, elements )
+import Test.QuickCheck.Extra
+    ( shrinkInterleaved )
+
+import qualified Data.Text as T
+
+--------------------------------------------------------------------------------
+-- Transaction hashes chosen from a small range (to allow collisions)
+--------------------------------------------------------------------------------
+
+genTxHashSmallRange :: Gen (Hash "Tx")
+genTxHashSmallRange = elements txHashes
+
+shrinkTxHashSmallRange :: Hash "Tx" -> [Hash "Tx"]
+shrinkTxHashSmallRange hash = filter (< hash) txHashes
```

Comment:
Small remark: I am not exactly sure what this is shrinking. Unless I am missing something, there's no way to shrink a hash.

Reply:
> Small remark: I am not exactly sure what this is shrinking. Unless I am missing something, there's no way to shrink a hash.

In the case of `genTxHashSmallRange`, hashes are initially chosen from a small range, so as to allow collisions:

```hs
txHashes :: [Hash "Tx"]
txHashes = mkTxHash <$> ['0' .. '7']
```

The shrinker doesn't shrink the hashes themselves: it merely returns a small list of hashes that are _less than_ the given hash in the available set.

```hs
> generate genTxHashSmallRange
> a = it
> toText a
  "4444444444444444444444444444444444444444444444444444444444444444"

> toText <$> shrinkTxHashSmallRange a
[ "0000000000000000000000000000000000000000000000000000000000000000"
, "1111111111111111111111111111111111111111111111111111111111111111"
, "2222222222222222222222222222222222222222222222222222222222222222"
, "3333333333333333333333333333333333333333333333333333333333333333"
]
```

In some cases, this can lead to simpler counterexamples.